libkbfs/semaphore_disk_limiter.go

```diff
@@ -0,0 +1,110 @@
+// Copyright 2017 Keybase Inc. All rights reserved.
+// Use of this source code is governed by a BSD
+// license that can be found in the LICENSE file.
+
+package libkbfs
+
+import (
+	"sync"
+
+	"github.com/keybase/kbfs/kbfssync"
+	"golang.org/x/net/context"
+)
+
+// semaphoreDiskLimiter is an implementation of diskLimiter that uses
+// a semaphore.
+type semaphoreDiskLimiter struct {
+	s                    *kbfssync.Semaphore
+	availableByteDivisor int64
+	maxByteLimit         int64
+
+	lock sync.RWMutex
+	// byteLimit is min(maxByteLimit,
+	// (availableBytes + journalBytes) / availableByteDivisor).
+	byteLimit      int64
+	journalBytes   int64
+	availableBytes uint64
+}
+
+var _ diskLimiter = (*semaphoreDiskLimiter)(nil)
+
+func newSemaphoreDiskLimiter(
+	availableByteDivisor, maxByteLimit int64) *semaphoreDiskLimiter {
+	if availableByteDivisor < 1 {
+		panic("availableByteDivisor must be >= 1")
+	}
+	if maxByteLimit <= 0 {
+		panic("maxByteLimit must be > 0")
+	}
+
+	s := kbfssync.NewSemaphore()
+	s.Release(maxByteLimit)
+	return &semaphoreDiskLimiter{
+		s:                    s,
+		availableByteDivisor: availableByteDivisor,
+		maxByteLimit:         maxByteLimit,
+
+		byteLimit: maxByteLimit,
+	}
+}
+
+func (s *semaphoreDiskLimiter) getByteLimit() int64 {
+	s.lock.RLock()
+	defer s.lock.RUnlock()
+	return s.byteLimit
+}
+
+func (s *semaphoreDiskLimiter) recalculateLimitsLocked() {
+	// TODO: Check for overflow.
+	dividedAvailableBytes := (s.availableBytes + uint64(s.journalBytes)) /
+		uint64(s.availableByteDivisor)
+	var newByteLimit int64
+	if dividedAvailableBytes > uint64(s.maxByteLimit) {
+		newByteLimit = s.maxByteLimit
+	} else {
+		newByteLimit = int64(dividedAvailableBytes)
+	}
+
+	oldByteLimit := s.byteLimit
+	s.byteLimit = newByteLimit
```

Comment:
We probably want to log the new limit whenever it changes (and thus have an `s.log` that's passed in on construction).